src/aws-healthomics-mcp-server/awslabs/aws_healthomics_mcp_server/models/s3.py

```diff
@@ -0,0 +1,396 @@
+        """Detailed string representation."""
+        return f'S3File(bucket="{self.bucket}", key="{self.key}")'
+
+
+# S3 File Utility Functions
+
+
+def create_s3_file_from_object(
+    bucket: str, s3_object: Dict[str, Any], tags: Optional[Dict[str, str]] = None
+) -> S3File:
+    """Create an S3File instance from an S3 object dictionary.
+
+    Args:
+        bucket: S3 bucket name
+        s3_object: S3 object dictionary from list_objects_v2 or similar
+        tags: Optional tags dictionary
+
+    Returns:
+        S3File instance
+    """
+    return S3File(
+        bucket=bucket,
+        key=s3_object['Key'],
+        size_bytes=s3_object.get('Size'),
+        last_modified=s3_object.get('LastModified'),
+        storage_class=s3_object.get('StorageClass'),
+        etag=s3_object.get('ETag', '').strip('"'),  # Remove quotes from ETag
+        tags=tags or {},
+    )
+
+
+def build_s3_uri(bucket: str, key: str) -> str:
+    """Build an S3 URI from bucket and key components.
+
+    Args:
+        bucket: S3 bucket name
+        key: S3 object key
+
+    Returns:
+        Complete S3 URI
+
+    Raises:
+        ValueError: If bucket or key is invalid
+    """
+    if not bucket:
+        raise ValueError('Bucket name cannot be empty')
+    if not key:
+        raise ValueError('Object key cannot be empty')
+
+    return f's3://{bucket}/{key}'
+
+
+def parse_s3_uri(uri: str) -> Tuple[str, str]:
+    """Parse an S3 URI into bucket and key components.
+
+    Args:
+        uri: S3 URI (e.g., 's3://bucket/path/to/file.txt')
+
+    Returns:
+        Tuple of (bucket, key)
+
+    Raises:
+        ValueError: If the URI format is invalid
+    """
+    if not uri.startswith('s3://'):
+        raise ValueError(f"Invalid S3 URI format: {uri}. Must start with 's3://'")
+
+    parsed = urlparse(uri)
+    bucket = parsed.netloc
+    key = parsed.path.lstrip('/')
+
+    if not bucket:
+        raise ValueError(f'Invalid S3 URI format: {uri}. Missing bucket name')
+
+    if not key:
+        raise ValueError(f'Invalid S3 URI format: {uri}. Missing object key')
+
+    return bucket, key
+
+
+def get_s3_file_associations(primary_file: S3File) -> List[S3File]:
+    """Get potential associated files for a primary S3 file based on naming conventions.
+
+    Args:
+        primary_file: Primary S3File to find associations for
+
+    Returns:
+        List of potential associated S3File instances
+
+    Note:
+        This function generates potential associations based on common patterns.
+        The actual existence of these files should be verified separately.
+    """
+    associations = []
+
+    # Common index file patterns
+    index_patterns = {
+        '.bam': ['.bam.bai', '.bai'],
+        '.cram': ['.cram.crai', '.crai'],
+        '.vcf': ['.vcf.tbi', '.tbi'],
+        '.vcf.gz': ['.vcf.gz.tbi', '.tbi'],
+        '.fasta': ['.fasta.fai', '.fai'],
+        '.fa': ['.fa.fai', '.fai'],
+        '.fna': ['.fna.fai', '.fai'],
+    }
```

Comment:
nit: Should we extract this out to a common enum for other usage? (I see similar references in `GenomicsFileType` and again for regex pattern matching.)